src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -315,14 +329,31 @@ public async Task<LSP.CodeAction> HandleRequestAsync(LSP.CodeAction codeAction,
                         }
 
                         var edits = textChanges.Select(tc => ProtocolConversions.TextChangeToTextEdit(tc, oldText)).ToArray();
-                        var documentIdentifier = new OptionalVersionedTextDocumentIdentifier { Uri = newTextDoc.GetURI() };
-                        textDocumentEdits.Add(new TextDocumentEdit { TextDocument = documentIdentifier, Edits = edits });
+
+                        if (edits.Length > 0)
+                        {
+                            var documentIdentifier = new OptionalVersionedTextDocumentIdentifier { Uri = newTextDoc.GetURI() };
+                            textDocumentEdits.Add(new TextDocumentEdit { TextDocument = documentIdentifier, Edits = edits });
+                        }
+
+                        // Rename
+                        if (oldTextDoc.State.Attributes.Name != newTextDoc.State.Name)
+                        {
+                            textDocumentEdits.Add(new RenameFile() { OldUri = oldTextDoc.GetUriFromName(), NewUri = newTextDoc.GetUriFromName() });
+                        }
```

Comment:
Does the order matter here relative to the text edits? This order seemse logical (we'd edit the file, then rename it) but if it does matter we probably should get a comment to that effect.

Reply:
> If resource operations are present clients need to execute the operations in the order in which they are provided.

It is listed in https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit
So I think it does matter. will add a comment